### DISCUSSION TITLE
XML Parsing Error and Questions

### DISCUSSION THREAD

**Discussion by gregwsil** [BLeeEZ/amperfy#365] [General]
This app kept me up last night! Not because I was having a bad time, but because i was so excited to test it out and finally bring my server to my iPhone and ditch Apple Music entirely! My idea is to delete the offline music from my phone synced via Apple Music (DRM free stuff)... then resync it down via Amperfy. Then I dont need to rely on plugging in my phone all the time to get new albums I download and such. This app looks like it solves that for me. +1

So, I have about 60GB / 10K songs in MP3 format. I have a server running via NextCloud AIO Music. I am able to log in to Amperfy, see the music on the server... all that jazz. However, I noticed that not all music was there. I checked the logs and it looks like it is quitting around 42% (somewhere around 4200 songs just being read from the server, not downloading).

Checking the logs in Nextcloud Music doesnt show any failures and such. When i check the logs in the Amperfy app, its showing a bunch of entries of "XML response could not be parsed". 

 I have tried rescans, deleting and reinstalling the app, etc to no avail. Any ideas on how to fix this and what troubleshooting steps could be? 

Also, question... If i add new music to my server, do i need to rescan the entire library, or will it pull down those changes? And if i previously sync my entire library, will i need to resync the entire library, or will it look for just modified changes?

**Comment by Schlaubischlump** [comment]
Which protocol are you using? NextCloud supports Ampache as well as Subsonic. I did not have any problems so far by using the Subsonic backend. When configuring Amperfy let it autodetect the protocol, then enter your nextcloud URL that ends with: 

`/index.php/apps/music/subsonic`

For me changes are detected automatically. No need to rescan.

**Reply by gregwsil** [reply]
I tried both Ampache and Subsonic APIs.

**Reply by Schlaubischlump** [reply]
Then I'm sadly out of ideas... Maybe BLeeEZ knows what is going on.

**Comment by gregwsil** [comment]
On a side note, as a test, i installed Navidrome on my server, imported the same folders and connected it to Amperfy. It looks like it is working fine!

Doesn't necessarily solve my issue, but a possible workaround for the time being while i test this new process. 

SUPER excited to get this to work if it does. 

@Schlaubischlump  - Do you know the answer to this....
Also, question... If i add new music to my server, do i need to rescan the entire library, or will it pull down those changes? And if i previously sync my entire library, will i need to resync the entire library, or will it look for just modified changes?

**Reply by Schlaubischlump** [reply]
Are there any special characters in the names of your music files that could be problematic? In this case it could be some weird behaviour where Navidrome encodes the characters, but Nextcloud does not. 

> Do you know the answer to this....

It all works automatically. My current setup is this: 
1. Point or move the `Music.app` library folder on your Mac to a folder on your Nextcloud. 
2. Point the Nextcloud Music app to the same folder
3. Setup Amperfy with the Subsonic API on my iPhone. 
4. Whenever I make any changes in the `Music.app` on my Mac (changing Metadata, adding files, deleting files) they automatically sync to Nextcloud. 
5. Amperfy automatically picks up any changes that occured in the Nextcloud, no need for manually rescanning anything.

**Reply by gregwsil** [reply]
Hmm. Do you happen to know what special characters? I am sure there are songs in my library with them, but probably only the following: !@#$%^&*()-+:",.'/
I'd have to pull a report to see exactly what doesnt. What's odd is: Nextcloud doesnt have an issue with the special characters, but maybe the API does? 

To the automatic piece, this is how i did it as well. The music folder stores everything in nextcloud. If there is an update, then the sync happens and it puts it in there. 
I just did a test and it works fine with navidrome, i was able to see a newly added file after closing and opening the app. 

Do you happen to know if the library is downloaded (not just synced, but downloaded for offline play), will it just look for changes in the library or does it wipe and redownload everything?

**Reply by Schlaubischlump** [reply]
Does the log contain more information on the parsing error? 

To the best of my knowledge there is no automatic download. I think you need to manually download music tracks in Amperfy. I don't know what happens if a track is already downloaded and you change it in on the server.

**Comment by gregwsil** [comment]
I have **good** news on the syncing piece. 

After uploading my entire library, I added a song as a test...
- Navidrome found the file and shortly after it was able to be seen in Amperfy. Cool!
- Clicking Download All Songs In Library did not successfully give me the download of the single track like I would have expected. 
- However, by clicking "Resync Library", the app shut down and I reopened. The cached songs were still there. 
- Then, clicking "Download All Songs In Library", the single track showed up and was cached to the unit. i.e., I did not have to redownload the entire library for one song. 

THIS IS A GAME CHANGER!

@BLeeEZ - Two suggestions for enhancements: 

**1 --** I shouldn't have to do a resync of the entire library (close down the app, refresh the cache, etc) to find a song that was added to the library. When I added the song and Navidrome picked it up, shortly after Amperfy saw that there was one added. By clicking Download All... I would have expected that it would have checked the server library against the local cache, saw the song wasn't there, updated the library accordingly, then cache it to the device. 

**2 --** When playing a song in offline mode, then turning offline mode off... Navidrome did not update the play count on the server. However, if I played the same song again from the app (with offline mode off), Navidrome then updated the play count to 2 (accounting for the one play that was offline). This should live in a queue somewhere waiting for server connection when offline mode is turned off to update ratings, playcounts, etc on the next connect. 

@Schlaubischlump -- I am going to try again using NextCloud vs Navidrome. Obviously Navidrome is the better way to go here, but having everything unified in one place is also comforting. If worst comes to worst, Navidrome can live on my server hosting the files out to the devices and update when requested. I will post more on this thread as I continue to test.

**Comment by BLeeEZ** [comment]
Hi @gregwsil,

1 — There is an option in „Settings->Library“ Auto Cache: Newest Songs. That should do what you requested.

2 — offline played songs are added to a persistent scrobble queue. This queue is synced to server when offline mode ended. This queue is only handling scrobbling. Mark favorite, rating and playlist changes are deactivated in offline mode.

@Schlaubischlump thank you for already answering and  managing issue/feature requests.

**Reply by gregwsil** [reply]
@BLeeEZ - 

1 -- You're right, that option is there and I completely missed it. Tested and it works great on app refresh. THANK YOU. 

2 -- Can you clarify "Mark favorite, rating and playlist changes are deactivated in offline mode.".... Are you saying that if I am in offline mode and I rate something, then go back online, that rating doesn't upload to the server on the same queue? 

Last question, and it's a dumb one, but I just want to confirm... Cache songs are preferred first over server songs right? Meaning, I am not offline and I select a song that is already cached on my device... I am playing the device copy, not the server copy, right? 

Thank you ALL! You have all been a huge help in understanding this wonderful piece of software.

**Comment by BLeeEZ** [comment]
2 -- rating and favourite should be deactivated everywhere so you can't change that when offline mode is active. I see that even if you are in offline mode the rating and favourite option is available in the preview menu. I will fix that. 

The cached files (local copy) will be used when available. If no cache is available the song/podcast episode will be streamed.

**Comment by BLeeEZ** [comment]
Favorite and rating fixed via dbc233ea63ef805b04751f13d96e9ebec2333c89.
